File: ios/TappyAI/Features/Music/UI/MusicLibraryViewModel.swift
import Foundation
import Combine

@MainActor
final class MusicLibraryViewModel: AppObservableObject {
    @AppPublished var tracks: [MusicTrack] = []
    @AppPublished var categories: [MusicCategory] = []
    @AppPublished var activeCategoryId: String?
    @AppPublished var loading = true
    @AppPublished var error: String?
    @AppPublished var hasMore = false
    @AppPublished var page = 0

    @AppPublished var searchQuery = ""
    @AppPublished var searchResults: [MusicTrack] = []
    @AppPublished var searchLoading = false
    @AppPublished var searchError: String?

    @AppPublished var previewingTrackId: String?

    private let service: MusicService
    private var searchDebounceTask: Task<Void, Never>?
    private let log = AppLogger.music

    var isSearching: Bool {
        !searchQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var displayTracks: [MusicTrack] {
        isSearching ? searchResults : tracks
    }

    var displayLoading: Bool {
        isSearching ? searchLoading : loading
    }

    var displayError: String? {
        isSearching ? searchError : error
    }

    init(service: MusicService) {
        self.service = service
    }

    // MARK: - Load categories

    func loadCategories() {
        Task {
            do {
                categories = try await service.getCategories()
            } catch {
                log.error("load music categories failed: \(error)")
            }
        }
    }

    // MARK: - Browse tracks (paginated)

    private var loadTask: Task<Void, Never>?

    func loadTracks() {
        loadTask?.cancel()
        loading = true
        error = nil
        page = 0

        loadTask = Task {
            do {
                let result = try await service.browseTracks(categoryId: activeCategoryId, page: 0)
                tracks = result.tracks
                hasMore = result.hasMore
                page = 1
            } catch {
                self.error = "Không thể tải danh sách nhạc"
                log.error("load music tracks failed: \(error)")
            }
            loading = false
        }
    }

    func loadMore() {
        guard !loading, hasMore else { return }
        loading = true

        Task {
            do {
                let result = try await service.browseTracks(categoryId: activeCategoryId, page: page)
                tracks.append(contentsOf: result.tracks)
                hasMore = result.hasMore
                page += 1
            } catch {
                log.error("load more music failed: \(error)")
            }
            loading = false
        }
    }

    // MARK: - Category selection

    func selectCategory(_ id: String?) {
        activeCategoryId = id
        loadTracks()
    }

    // MARK: - Search (300ms debounce, matching Web)

    func updateSearch(_ query: String) {
        searchQuery = query
        searchDebounceTask?.cancel()

        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            searchResults = []
            searchLoading = false
            searchError = nil
            return
        }

        searchLoading = true
        searchError = nil

        searchDebounceTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            do {
                let result = try await service.searchTracks(query: trimmed)
                guard !Task.isCancelled else { return }
                searchResults = result.tracks
            } catch {
                guard !Task.isCancelled else { return }
                searchError = "Không thể tìm nhạc"
                log.error("music search failed: \(error)")
            }
            searchLoading = false
        }
    }

    // MARK: - Preview toggle

    func togglePreview(_ track: MusicTrack) {
        if previewingTrackId == track.id {
            previewingTrackId = nil
        } else {
            previewingTrackId = track.id
        }
    }

    func stopPreview() {
        previewingTrackId = nil
    }
}

File: android/app/src/main/java/com/tappyai/app/music/MusicLibraryViewModel.kt
package com.tappyai.app.music

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tappyai.app.music.data.MusicErrorMessages
import com.tappyai.app.music.data.MusicRepository
import com.tappyai.core.common.UiState
import com.tappyai.core.logging.LoggerProvider
import com.tappyai.core.network.NetworkResult
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * State for the Music Library — mirrors the web `/music` page: browse by category (paginated,
 * `GET /api/music/tracks`) or search across everything (debounced, `GET /api/music/tracks/
 * search`), which hides the category tabs while active. Both endpoints are public (no sign-in
 * required), matching the web.
 */
@HiltViewModel
class MusicLibraryViewModel @Inject constructor(
    private val repository: MusicRepository,
    private val logger: LoggerProvider,
    private val musicErrorMessages: MusicErrorMessages,
) : ViewModel() {

    var categories by mutableStateOf<List<MusicCategory>>(emptyList())
        private set

    var query by mutableStateOf("")
        private set

    /** Selected category id, or `null` for the "All" tab. */
    var selectedCategoryId: String? by mutableStateOf(null)
        private set

    var tracksState by mutableStateOf<UiState<List<MusicTrack>>>(UiState.Loading)
        private set

    var isLoadingMore by mutableStateOf(false)
        private set

    val isSearching: Boolean
        get() = query.isNotBlank()

    private var currentPage = 0
    private var hasMore = false
    private var loadJob: Job? = null
    private var searchDebounceJob: Job? = null

    init {
        loadCategories()
        loadTracks(reset = true)
    }

    private fun loadCategories() {
        viewModelScope.launch {
            when (val result = repository.getCategories()) {
                is NetworkResult.Success -> categories = result.data
                is NetworkResult.Error -> logger.e(TAG, "Categories load failed: ${result.error}")
            }
        }
    }

    fun onQueryChange(value: String) {
        query = value
        searchDebounceJob?.cancel()
        searchDebounceJob = viewModelScope.launch {
            delay(SEARCH_DEBOUNCE_MS)
            loadTracks(reset = true)
        }
    }

    fun onSelectCategory(categoryId: String?) {
        if (selectedCategoryId == categoryId) return
        selectedCategoryId = categoryId
        loadTracks(reset = true)
    }

    fun retry() = loadTracks(reset = true)

    /** Call when the list's last visible row nears the end — appends the next page, or no-ops if
     *  a load is already in flight or the backend already reported no more pages. */
    fun loadMore() {
        if (isLoadingMore || !hasMore) return
        loadTracks(reset = false)
    }

    private fun loadTracks(reset: Boolean) {
        loadJob?.cancel()
        if (reset) {
            currentPage = 0
            tracksState = UiState.Loading
        } else {
            isLoadingMore = true
        }
        loadJob = viewModelScope.launch {
            val result = if (isSearching) {
                repository.searchTracks(query.trim(), currentPage)
            } else {
                repository.getTracks(selectedCategoryId, currentPage)
            }
            when (result) {
                is NetworkResult.Success -> {
                    val page = result.data
                    hasMore = page.hasMore
                    currentPage = page.page + 1
                    val existing = (tracksState as? UiState.Success)?.data ?: emptyList()
                    val combined = if (reset) page.tracks else existing + page.tracks
                    tracksState = if (combined.isEmpty()) UiState.Empty else UiState.Success(combined)
                }
                is NetworkResult.Error -> {
                    logger.e(TAG, "Tracks load failed: ${result.error}")
                    // A failed "load more" silently drops — the existing list stays visible, matching
                    // standard infinite-scroll UX (only a fresh/reset load surfaces a hard error state).
                    if (reset) {
                        tracksState = UiState.Error(musicErrorMessages.toUserMessage(result.error))
                    }
                }
            }
            isLoadingMore = false
        }
    }

    private companion object {
        const val TAG = "MusicLibraryViewModel"
        const val SEARCH_DEBOUNCE_MS = 400L
    }
}
